### PR TITLE
[AI] Better openAi prompt

### LINKED ISSUE #28550: [Jetpack AI] posts without a title
Once https://github.com/Automattic/jetpack/pull/28483 is merged, we have to allow more cases of generating a prompt, when no title is chosen:

- only content
- content + categories
- only categories

The only generation we should disallow when the post is absolutely empty

### PR DESCRIPTION
**Pull request by millerf**
Depending on #28483 to be merged

Fixes #28550

## Proposed changes:
* Handles more post cases when creating the OpenAI prompt
* Slightly Refactor it
* Only display an error when the post is totally empty
* Started to add unit tests

### Other information:
 [X ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

No
## Testing instructions:
* Deploy on WPCOM
* Sandbox your favorite site
* Go to `{YOUR_SITE}/wp-admin/post-new.php`
* Add an AI paragraph block to the empty page and you should see
Screenshot 2023-01-25 at 14 48 23

* Entering a title, some content, a category or a tag should ask you to retry
* You can now prompt for each case and see the different prompts in the network console
* All types of prompts are unit-tested in `extensions/blocks/ai-paragraph/test/edit.js`

## E2E tests

Run
```
 pnpm test-extensions extensions/blocks/ai-paragraph/test/edit.js
```

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js [hunk 1]
```diff
@@ -193,20 +186,13 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
...
+		setNeedsMoreCharacters( false );
 		setIsLoadingCompletion( true );
 
 		const data = {
-			content: createPrompt( currentPostTitle, contentBefore, categoryNames ),
-			post_id: postId,
+			content: createPrompt( currentPostTitle, contentBefore, categoryNames, tagNames ),
 		};
-
-		tracks.recordEvent( 'jetpack_ai_gpt3_completion', {
```
Comment (artpi) [inline projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:206]: Please dont remove the track s event

File: projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import './editor.scss';
 
-import { useAnalytics } from '@automattic/jetpack-shared-extension-utils';
```
Comment (artpi) [inline projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js]: Dont remove anaalytics

**Review by artpi — CHANGES_REQUESTED** [review]
- I like the tests
- This seems to remove tracks events
- I am not sure if we want to proceed with this angle or do we want to move prompt engineering to the backend as in https://github.com/Automattic/jetpack/issues/28569 ?

File: projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js (hunk 1 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:206]: It has to have been an issue when I merged trunk into it. I didn't remove that on purpose...

**Comment by millerf** [comment]
> I am not sure if we want to proceed with this angle or do we want to move prompt engineering to the backend as in https://github.com/Automattic/jetpack/issues/28569 ?

Those 2 are not incompatible: that PR might be a nice middle-step as the prompt logic is already written. We can then easily "port" the logic (and tests) to the back-end.... What do you think @artpi ?

**Comment by artpi** [comment]
> Those 2 are not incompatible: that PR might be a nice middle-step as the prompt logic is already written. We can then easily "port" the logic (and tests) to the back-end....

Do you mean port it to PHP? Sure.

But I think this logic should be different anyway, as the solution to the output stopping mid-sentence is to write a different prompt.

File: projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js [hunk 3]
```diff
@@ -9,14 +9,16 @@ import { useState, RawHTML, useEffect } from '@wordpress/element';
 import { sprintf, __ } from '@wordpress/i18n';
 import { name as aiParagraphBlockName } from './index';
 
-// Minimum number of characters needed in the prompt to send it to the backend
-const NUMBER_OF_CHARACTERS_NEEDED = 36;
-
 // Maximum number of characters we send from the content
-const MAXIMUM_NUMBER_OF_CHARACTERS_SENT_FROM_CONTENT = 240;
+export const MAXIMUM_NUMBER_OF_CHARACTERS_SENT_FROM_CONTENT = 240;
```
Comment (artpi) [inline projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:13]: Let's bump it maybe to 512?

**Review by artpi — APPROVED** [review]
This works well!
It will be easier to tweak the prompts with these tests.

**Review by jeherve — APPROVED** [review]

**Pull request merged**